File: src/rule.py
# -*- coding: utf-8 -*-
#-------------------------------------------------------------------------------
"""
    @license
    Copyright (c) Daniel Pauli <dapaulid@gmail.com>

    This source code is licensed under the MIT license found in the
    LICENSE file in the root directory of this source tree.
"""
#-------------------------------------------------------------------------------

#-------------------------------------------------------------------------------
# imports
#-------------------------------------------------------------------------------
#
import np
import json
import os
import functools
import math

#-------------------------------------------------------------------------------
# constants
#-------------------------------------------------------------------------------
#
## convolution kernel used to determine index into rule array for each cell neighborhood
NEIGH_KERNEL = np.array(
   [[4, 3, 2],
    [5, 0, 1],
    [6, 7, 8]]
)


## the base64 alphabet used for rule encoding. 
## it is chosen to be URL friendly and maintain 0 for zero bits. 
B64_CHARS = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ.-"
B64_IDX = {}
for i in range(len(B64_CHARS)):
	B64_IDX[B64_CHARS[i]] = i
# end for
B64_MASK  = 0x3F
B64_SHIFT = 6

#-------------------------------------------------------------------------------
# class definition
#-------------------------------------------------------------------------------
#
class Rule:

	@staticmethod
	def get_array_size(states):
		return states ** NEIGH_KERNEL.size
	# end function

	#---------------------------------------------------------------------------
    ## constructor
	def __init__(self, array):
		self.states = int(len(array) ** (1.0/NEIGH_KERNEL.size))
		self.size = Rule.get_array_size(self.states)
		self.array = np.array(array)
		self.kernel = np.power(self.states, NEIGH_KERNEL)
	# end function

    #---------------------------------------------------------------------------
    ## applies this rule to the given two-dimensional array
	def apply(self, cells):
		# TODO optimized convolve function?
		# http://blog.rtwilson.com/convolution-in-python-which-function-to-use/#:~:text=convolve%20is%20about%20twice%20as,convolve2d.&text=Using%20a%20random%208000%20x,a%20loop%20of%20various%20convolutions.
		idx = np.convolve(cells, self.kernel, mode='wrap')
		# determine next state by applying the rules
		new_cells = self.array[idx] # same as self.arr.take(idx), but seems faster
		return new_cells, idx
	# end function

    #---------------------------------------------------------------------------
    ## returns the string representation (base64) of this rule
	def encode(self):
		encoded = ""
		num = functools.reduce(lambda acc, x: acc * self.states + int(x), self.array[::-1], 0)
		for i in range(Rule.get_base64_size(self.states)):
			encoded += B64_CHARS[num & B64_MASK]
			num >>= B64_SHIFT
		# end for
		return encoded		
	# end function

	@staticmethod
	def decode(str):
		# determine number of states
		states = None
		for i in range(2, 256):
			if Rule.get_base64_size(i) == len(str):
				states = i
				break
			# end if
		# end if
		num = 0
		for c in str[::-1]:
			num <<= B64_SHIFT
			num += B64_IDX[c]
		# end for
		n = states ** 9
		array = [None] * n
		for i in range(n):
			array[i] = num % states
			num //= states
		# end for
		print(array)
		return Rule(array)
	# end function	

	def get_count(self):
		return np.bincount(self.array, minlength=self.states) / self.size	
	# end function

	def __str__(self):
		# TODO output base64 representation?
		return "(" + ", ".join(["%.3f" % x for x in self.get_count()]) + ")"
	# end function

    #---------------------------------------------------------------------------
    ## returns the file name of the rule file
	@staticmethod
	def get_rule_file(name):
		return os.path.join("data/rules", name + ".json")
	# end function

    #---------------------------------------------------------------------------
    ## saves the rule with the given name
	def save(self, name):
		state = {
			'array': self.array.tolist()
		}
		with open(Rule.get_rule_file(name), 'w') as f:
			json.dump(state, f)
		# end with
	# end function

    #---------------------------------------------------------------------------
    ## loads the rule with the given name
	@staticmethod
	def load(name):
		with open(Rule.get_rule_file(name), 'r') as f:
			state = json.load(f)
		# end with
		return Rule(state['array'])
	# end function

    #---------------------------------------------------------------------------
    ## constructs a rule from a given (big) integer
	@staticmethod
	def from_int(value, states):
		n = Rule.get_array_size(states)
		array = [0] * n
		for i in range(n):
			array[i] = value % states
			value //= states
		# end for
		return Rule(array)
	# end function	

	@staticmethod
	def get_base64_size(states):
		return math.ceil(states**9 * math.log(states)/math.log(64))
	# end function

# end class

#-------------------------------------------------------------------------------
# end of file

File: web/client/rule.js

/** 
 * the base64 alphabet used for rule encoding. 
 * it is chosen to be URL friendly and maintain 0 for zero bits. 
 */
const B64_CHARS = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ.-";
const B64_IDX = {}
for (let i = 0; i < B64_CHARS.length; i++) {
	B64_IDX[B64_CHARS[i]] = i;
}

// constants for BigInt operations
const B64_MASK  = 0x3Fn
const B64_SHIFT = 6n;

const NEIGH_SIZE = 9; // size of a 3x3 Moore neighborhood including center

class Rule {

	constructor(array) {
		this.array = array;
		this.states = Math.pow(this.array.length, 1/NEIGH_SIZE);
	}

	encode() {
		const radix = BigInt(this.states);
		let num = this.array.reduce((acc, x) => acc * radix + BigInt(x), BigInt(0));
		const len = getRuleSize(this.states);
		let encoded = "";
		for (let i = 0; i < len; i++) {
			encoded += B64_CHARS[Number(num & B64_MASK)];
			num >>= B64_SHIFT;
		}
		return encoded;
	}

	static decode(str) {
		// determine number of states
		let states = null;
		for (let i = 2; i < 256; i++) {
			if (getRuleSize(i) == str.length) {
				states = i;
				break;
			}
		}
		let num = BigInt(0);
		for (let i = str.length-1; i >= 0; i--) {
			num <<= B64_SHIFT;
			const val = B64_IDX[str[i]];
			num += BigInt(val);
		}
		const radix = BigInt(states);
		const n = Math.pow(states, NEIGH_SIZE);
		let array = [];
		for (let i = 0; i < n; i++) {
			array[n-i-1] = Number(num % radix);
			num /= radix;
		}
		return new Rule(array);
	
	}

	static random(states) {
		if (states != 2) {
			throw Error("Only 2 states supported so far");
		}
		const n = states**NEIGH_SIZE;
		const t = Math.random();
		let array = []
		for (let i = 0; i < n; i++) {
			array[i] = Math.random() <= t ? 0 : 1;
		}
		return new Rule(array);
	}

	static generate(states, func) {
		let array = []
		let neigh = zeros(NEIGH_SIZE+1); // add one for sentinel
		while (!neigh[NEIGH_SIZE]) {

			// determine state of center cell
			let cell = neigh[0];
			
			// count cells for each states
			// start from 1 to exclude center
			let counts = zeros(states);
			for (let i = 1; i < NEIGH_SIZE; i++) {
				counts[neigh[i]]++;
			}

			// determine transition for this configuration
			let next_cell = func(cell, counts);
			if (!Number.isInteger(next_cell)) {
				throw TypeError("returned state must be an integer");
			}
			if ((next_cell < 0) || (next_cell >= states)) {
				throw RangeError("returned state must be between 0 and " + (states-1));
			}
			array.push(next_cell);

			// increment with overflow
			for (let i = 0; i <= NEIGH_SIZE; i++) {
				neigh[i] = (neigh[i] + 1) % states;
				if (neigh[i] != 0) {
					break;
				}
			}
		}
		return new Rule(array);
	}

	get length() {
		return this.array.length;
	}
}

function getRuleSize(states) {
    return Math.ceil(Math.pow(states, 9) * Math.log(states)/Math.log(64));
}

function zeros(n) {
	return new Array(n).fill(0);
}